keps/wg-lts/20200122-kubernetes-yearly-support-period.md

```diff
@@ -0,0 +1,313 @@
+  * Ideally, the latest kubectl would support speaking to all supported apiservers (+0/-3 minor versions).
+  * Implementation should include adding a test to cover the added skew variations, have it release informing, watch for any errors, fix them, promote the test to release blocking.  This is already a gap today and can be treated as a project need orthogonal to this KEP.
+
+### Test/release process changes
+
+* Staff patch releases for one more branch for ~3-4 more months
+  * Feedback from SIG Release received, see [SIG Release meeting discussion](https://docs.google.com/document/d/1Fu6HxXQu8wl6TwloGUEOXVzZ1rwZ72IAhglnaAMCPqA/edit#bookmark=id.mi11nk75iohl).
+* Maintain CI for one more branch for ~3-4 more months
+  * Modifying test job and infrastructure patterns may need deprecated on a slower schedule or backwards compatibility with older style tests in older branches may need maintained longer.  This is a problem already today (e.g.: bootstrap.py went out of support, but was still in use in release-1.14 branch CI during the end of that branch’s lifetime).
+* Formalize turn-down timing of CI/ for oldest release and patch acceptance criteria for release branch:
+  * Historically this trailing period outside of support was squishy ("wait a little while, cut a final patch release, turn down jobs").  It is currently specified that oldest release branch CI support is turned off at the time when the newest, under-development release branch is created.  This happens at the first ‘beta’ release, e.g.: ‘1.16.0-beta.0’.  Due to an implementation detail where branches are referred to by “dev”, “beta”, “stable1”, “stable2”, “stable3” there is a rotation of marker variables at the point of new branch creation.  This is an automated process, run by the SIG Release subproject Release Engineering branch manager, and can be adjusted as needed.  But while CI is running if a CVE or upgrade issue is encountered and resolution is able to be backported the community has intended to do so.  In practice this has not been required.
+  * This KEP proposes the rotation happen 2 months after the 1 year period, i.e.: 12+2 months patch ability.  This constitutes 12 months of normal patch releases, plus a two month grace period for end users to plan and upgrade their clusters to a supported version. During this special 2-month grace period only critical security and upgrade fixes are acceptable on the branch.  I.e.:
+    * Support Policy during the first 12 months.
+      * Same as the current 9 months policy.
+    * Support Policy during the final +2 months.
+      * Only critical security patches and upgrade blocking fixes, i.e.:
+        * CVE assigned by Product Security Committee initiated to release branch by Product Security Committee
+        * Cherry-pick of upgrade scenario bug fix approved by owning SIG and Patch Release Team
+* Formalize 3rd party dependency update policy (e.g.: golang, but also others too)
```

Comment:
This point should be emphasized, as it would require upgrading and backporting non-critical fixes to ensure that security fixes also go back to previous releases. This could impact our stability/scalability guarantees.

This also includes supporting bazel fixes back, and bazel's interactions with golang, rules_go, etc.

This is more complex than it looks on the surface. cc: @liggitt

Reply:
+1

Reply:
I think we can reduce this potentially in the future, but right now this can get rather expensive and is somewhat understaffed, mostly done in highly active peoples' free time between other obligations.

Reply:
I'm not super aware of the Bazel project operationally or how that impacts us here.  They claim semver and "We generally avoid making backwards-incompatible changes."  What's our project track record on needing to backport bazel fixes to specific k/k release-X.YY branches?  Do we tend to carry patches on old versions versus move to a newer bazel?